torch/_inductor/codegen/triton.py

```diff
@@ -915,18 +922,22 @@ def indexing(
 
         if (need_dense and not have_dense) or isinstance(index, sympy.Integer):
             if copy_shape:
-                index_str = f"{index_str} + tl.zeros({copy_shape}.shape, tl.int32)"
+                index_str = (
+                    f"{index_str} + tl.zeros({copy_shape}.shape, {self.index_dtype})"
```

Comment:
I'm a bit concerned about `ops.index_expr` using 32-bit values in calculations that overflow, but I suppose a lot of lowerings currently hard-code 32-bit dtypes so are probably broken anyway.

Reply:
Working in the indirect indexing stuff, I've seen that plenty of lowerings hardcode that. It'd be nice at some point to reduce the number of lowerings so that these fixes affect more functions directly, without having to modify the lowerings.

Reply:
triton's `index_expr` overload ignores dtype anyway, for cpp it's a problem